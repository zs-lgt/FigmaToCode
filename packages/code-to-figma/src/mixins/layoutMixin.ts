import { LayoutProps } from '../types';
import { isValidSize } from '../helpers/isValidSize';

export const layoutMixin = (node: LayoutMixin & BaseNode) => (props: LayoutProps & { preventResizing?: boolean }) => {
    if (props.preventResizing) {
        return;
    }
    if (props.relativeTransform) {
        node.relativeTransform = props.relativeTransform;
    }
    if (typeof props.x === 'number') {
        node.x = props.x;
    }
    if (typeof props.y === 'number') {
        node.y = props.y;
    }
    if (typeof props.rotation === 'number') {
        node.rotation = props.rotation;
    }
    if ((isValidSize(props.width) || isValidSize(props.height)) && node.type !== 'LINE') {
        if (props.isWithoutConstraints) {
            node.resizeWithoutConstraints(props.width, props.height);
        } else {
            node.resize(
                isValidSize(props.width) ? props.width : node.width,
                isValidSize(props.height) ? props.height : node.height
            );
        }
    }

    if (isValidSize(props.width) && node.type === 'LINE') {
        if (props.isWithoutConstraints) {
            node.resizeWithoutConstraints(props.width, 0);
        } else {
            node.resize(props.width, 0);
        }
    }

    node.layoutAlign = props.layoutAlign || 'INHERIT';
    node.layoutGrow = props.layoutGrow || 0;
};
