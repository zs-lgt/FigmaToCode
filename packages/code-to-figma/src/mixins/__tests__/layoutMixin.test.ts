import { layoutMixin } from '../layoutMixin';
import { createFigma } from 'figma-api-stub';

describe('layoutMixin', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('resizing', () => {
        const resize = jest.fn((width, height) => ({ width, height }));
        layoutMixin({ type: 'RECTANGLE', resize } as any)({ width: 500, height: 200 });
        expect(resize.mock.results[0].value).toMatchSnapshot();
    });

    it('resizing without height', () => {
        const resize = jest.fn((width, height) => ({ width, height }));
        layoutMixin({ type: 'RECTANGLE', resize } as any)({ width: 500 });
        expect(resize.mock.calls.length).toBe(1);
    });

    it('resizing Line', () => {
        const resize = jest.fn((width, height) => ({ width, height }));
        layoutMixin({ type: 'LINE', resize } as any)({ width: 500 });
        expect(resize.mock.results[0].value).toMatchSnapshot();
    });

    it('apply x=0', () => {
        const node = { type: 'RECTANGLE', x: 200 } as any;
        layoutMixin(node)({ x: 0 });
        expect(node.x).toEqual(0);
    });

    it('apply y=0', () => {
        const node = { type: 'RECTANGLE', y: 100 } as any;
        layoutMixin(node)({ y: 0 });
        expect(node.y).toEqual(0);
    });

    it('resizing with zero width', () => {
        const node = figma.createRectangle();
        // @ts-ignore
        node.width = 100;
        // @ts-ignore
        node.height = 100;
        layoutMixin(node)({ width: 0, height: 10 });
        expect(node.width).toEqual(100);
        expect(node.height).toEqual(10);
    });

    it('resizing with negative width', () => {
        const node = figma.createRectangle();
        // @ts-ignore
        node.width = 100;
        // @ts-ignore
        node.height = 100;
        layoutMixin(node)({ width: -10, height: 10 });
        expect(node.width).toEqual(100);
        expect(node.height).toEqual(10);
    });

    it('resizing with zero height', () => {
        const node = figma.createRectangle();
        // @ts-ignore
        node.width = 100;
        // @ts-ignore
        node.height = 100;
        layoutMixin(node)({ width: 10, height: 0 });
        expect(node.width).toEqual(10);
        expect(node.height).toEqual(100);
    });

    it('resizing with negative height', () => {
        const node = figma.createRectangle();
        // @ts-ignore
        node.width = 100;
        // @ts-ignore
        node.height = 100;
        layoutMixin(node)({ width: 10, height: -10 });
        expect(node.width).toEqual(10);
        expect(node.height).toEqual(100);
    });

    it('setup layoutAlign', () => {
        const node = figma.createRectangle();
        layoutMixin(node)({ layoutAlign: 'CENTER' });
        expect(node.layoutAlign).toEqual('CENTER');
    });
});
