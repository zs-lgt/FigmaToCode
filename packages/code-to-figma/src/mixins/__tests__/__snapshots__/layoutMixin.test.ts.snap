// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layoutMixin resizing 1`] = `
Object {
  "height": 200,
  "width": 500,
}
`;

exports[`layoutMixin resizing Line 1`] = `
Object {
  "height": 0,
  "width": 500,
}
`;
